router/core/operation_parser.go

```diff
@@ -0,0 +1,159 @@
+package core
+
+import (
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/buger/jsonparser"
+	"github.com/cespare/xxhash/v2"
+	"github.com/wundergraph/cosmo/router/internal/pool"
+	"github.com/wundergraph/cosmo/router/internal/unsafebytes"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/ast"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/astnormalization"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/astparser"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/astprinter"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/operationreport"
+)
+
+type ParsedOperation struct {
+	// ID represents a unique-ish ID for the operation calculated by hashing
+	// its normalized representation and its variables
+	ID uint64
+	// Name is the operation name, if any
+	Name string
+	// Type is a string representing the operation type. One of
+	// "query", "mutation", "subscription"
+	Type string
+
+	// Query is the "query" field value in the JSON payload
+	Query string
+	// Variables in the "variables" field value in the JSON payload
+	Variables []byte
+	// NormalizedRepresentation is the normalized representation of the operation
+	// as a string. This is provided for modules to be able to access the
+	// operation.
+	NormalizedRepresentation string
+}
+
+type OperationParser struct {
+	executor     *Executor
+	documentPool *sync.Pool
+}
+
+func NewOperationParser(executor *Executor) *OperationParser {
+	return &OperationParser{
+		executor: executor,
+		documentPool: &sync.Pool{
+			New: func() interface{} {
+				return ast.NewSmallDocument()
+			},
+		},
+	}
+}
+
+func (p *OperationParser) ParseReader(r io.Reader) (*ParsedOperation, error) {
+	buf := pool.GetBytesBuffer()
+	defer pool.PutBytesBuffer(buf)
+	if _, err := io.Copy(buf, r); err != nil {
+		return nil, fmt.Errorf("failed to read request body: %w", err)
+	}
+	body := buf.Bytes()
+	return p.Parse(body)
+}
+
+func (p *OperationParser) Parse(body []byte) (*ParsedOperation, error) {
+	requestQuery, _ := jsonparser.GetString(body, "query")
```

Comment:
During the merging of ["limit request body size"](https://github.com/wundergraph/cosmo/pull/185/commits/e2f3c685fc1240a6a5f37e95d7fa20afb2f54745#diff-0c6c16945a1e9a216631c6bd16bcfbe9c07a6c96b1cf9b3aa93385436fab41e8R65) you forgot to move the logic to this place to respect the limit for WS as well.

Reply:
I'd rather have the upper layer do the request limiting, to provide better context for the errors. I'm now passing the limit to the websocket handler and producing an error there.